### ISSUE TITLE
Docs Meeting Notes: Mixins Reference Page

### ISSUE THREAD

**Issue by orta** [microsoft/TypeScript#39118]
> - **Date:** June 17th, 2020
> - **Agenda**
>   1. Overview of mixin docs today `5min`
>   2. Discussion of the goals for mixin support, and how we want to document it `10min`
>   3. Shortcomings worth documenting in current mixin implementation `10min`
>   4. Open discussion: `~`

## Mixin Docs Today

- [Handbook: Mixin docs](https://www.staging-typescript.org/docs/handbook/mixins.html)
- [Playground Sample: Mixin docs](https://www.staging-typescript.org/play?q=423#example/mixins)
- [Release notes](https://www.staging-typescript.org/docs/handbook/release-notes/typescript-2-2.html)

Some folks have looked to improve it, and add their own docs:

- [PR updating to 'ES2017'](https://github.com/microsoft/TypeScript-Handbook/pull/918/files)
- [Recommendations from @DustinWehr](https://github.com/microsoft/TypeScript-Handbook/issues/15)
- [Marius Schultz - Mixins](https://mariusschulz.com/blog/mixin-classes-in-typescript) (this is what the playground is based on)
- ['Real mixins with JS classes'](http://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/)

## Mixin goals

- Show working examples
- Show work-arounds for current big issues

#### Today, how do we recommend it

- Currently 2 different implementations strategies:
  - handbook: Use isolated `class`es. Then use `interface` merging + a runtime function to create - [play](https://www.staging-typescript.org/play?#code/PTAEBEEsGcAcHtoEMBGAbApqAspAHpAHYBQAxmktNBDAsulgN7GiugxRyIYAmAXKBTx4mJIVABeUADMkaaBgDcxFmx61uACgCUoZmwOgALgAsYAOg4aFPScYBOAVyWrWAX2IfiIUAEFSRpAAbkhGqJg4+ERkFFR+AcGh4Uyu7ND+gUEYAkIiGGJ2svIuqUgJIUYYOnqpBqTwhNB55mjwAOaaAEQZiZU8ndq1bKYWMD1ZdkZOLgYeBjz55aFVuvqGrCPQlunlWFJFCsqznirklNQAygC2SPZGAPIoAFYYATUGRJX2ZUbVa+ubcw-XorI5sDxeT4YeyyUhYa63B7PV5GUAYPCVQg8aicOjJAA08UySQYeg8SFgsDQAE9cARGpoEXdHi8AoSANq4xAEom9ZIAXW0ymImFR0BuzOedkIGAA7qAmUjWb8hcRxYiWZZCF8fjoVCADYajcaTaajd4wABJcTU+COeygJzayBXLBoSAob72amgJqu2UmaEYC1m0Nh00qaSOQgJBqgClU2lRBkLezBXgAYSM8HsAjE1MJKEoGCzOegecI1PZgvebCLClL9i20hzAFEyiZNPWS9mHRIAHy19YslHmNoYB6ywgABXs8Fg0KM1IAckhXdAu8XG+ZYHPs0uF9pzC37O3SJ3CGu9oP-us2COAuYFtIiBhZ-PF9TNKn0zxt7v4H3akF0JS9XUJB8jDHCd7ind8FzualwAwaBSDTWBe03Btex3PdAOAjBQKvbQAEJVTvUA3HI8FVQ8IA)
  - playground: (mixin factory pattern) `class`es use `extends` with their list of deps - [play](https://www.staging-typescript.org/play?#code/C4TwDgpgBAwg9gOwM7AE4FcDGwCGAjAGwgB4AVKAXigG8BfAPkqgQgHcoAKAOh51QHMkUAFxQcCEAG0AugEpKjUgCglmAjiRCAynAC20OioBm6BNgCWiKAFlzAD3MIAjGSgQ7wCAgAmQ+MjQsXEIIeg4AIQ1oUVJ5aiUoRKhUCGB0VAQoNQ0hAHlgAAsIVGs4bwgCNw8vXyhIpAMEpOb9QrKnDjim5p6sxCQ4Ii4COH4OACJUcShWgvbx2W6e2iWVlaUTM2BLTNsHBAAmV3dPHz9+wOx8IjD66KhYmiWUtIys9U0oABFzIyNriClcqVE41IR3J69RKzMoHTo0FbNNYqbKfUgQFBVU61PaOOG45wcHT6WRdZoDfScMlQwrmJBcGHeDqLJFKdZEYBQOB4ABWTBY7HRKE6Sm5PK4FIgIqAA)

- There is one library which covers a lot of mixin stuff: [ts-mixer](https://github.com/tannerntannern/ts-mixer) (no need to read it, it's mostly the `extends` pattern) - probably worth a sentence

- ~~The 4.0 improvements to tuples could improve the pattern [...maybe?](https://github.com/microsoft/TypeScript/issues/5453#issuecomment-154040685)~~

#### How to describe short-comings

- [Mixins with Generics](https://github.com/microsoft/TypeScript/issues/26154) - [play](https://www.staging-typescript.org/play?strictPropertyInitialization=false&ssl=1&ssc=1&pln=37&pc=1#code/PTAEFkEMGsFMGdSVAYwPYDt4BcBOBXFbNXSAIwBtYAoASw21lwDNIVZQBhTHAokgDwAVUAF4kGAJ4A+ULAAejDABNEAMXwYitTKADe1UKAywA7gAoAdNci4A5vABcEyQG0AugEpnQgNzUAX2pqZk1tXW4GSHomcFp5emFwNGVYCmlzMkh4WE4KbKcuHjxCYlwBSClZcQB5MgArWCJPfUNQXFhsfFwMVHz4RAUlVVAsnLyC1qMjAFsUtJ9k1Ip-IwD-ILoGJlZ2CHmKAEEp0GY0NGdeejsN4PpGFjYOJbSAIROs3Eu8a9vqEFAADk0NhTrBsrRKBwyE1IPgcqBJGh8KhKgByUEAC0gADcONhMRw7LATLhaChQPRQATIKCAA5oeC0bA6DDUFD9RCRbDRUlJA6yIYkkbc3mxeKJIQvdLmFp6TbBdBYUFKnkxXBqEjS47iEymIpRdUCbUZTz+VVijVag6HSxzZZifQE2jwACS8EOUmcaM+aPW7OKqEwatJmtw0veurMBpDTGNB1epvNwctYYjdoOjr01MxLvdhwo8AuoB9kAAXmjQP7-mAAOokaC2ZEqSmIYhyeR0ijk5kUSSoDq0jjIDmTM64RHIiczCW9SCEyDKGugNDMHP42gzGijgYx1MkY5ClRclPquIJDDx5aHU36Kvszl79Vh95HkWn0nnxIR2-yxWB0wGybTRlFFZ9rWvR09SfUMD1lfxANwRtcGbUCPyYNMbQzB1xGzM5iwAIgI+8gA)

  Roughly the static part of a class is a singleton (and doesn't vary), so you can't do something like `class A<T> extends B<T>`. The workaround is to declare the class independently.
  
  You can move the position of the [generics - example](https://www.staging-typescript.org/play?strictPropertyInitialization=false#code/PTAEFkEMGsFMGdSVAYwPYDt4BcBOBXFbNXSAIwBtYAoASw21lwDNIVZQBhTHAokgDwAVUAF4kGAJ4A+ULAAejDABNEAMXwYitTKADe1UKAywA7gAoAdNci4A5vABcEyQG0AugEpnQgNzUAX2pqEFAASWxUSAwAckiyDnx4WGVQZhJQHEhsWhQAGlAyfEiUaMLE5NTdbAALDnoTXFQKSHgnamZNbV1uBkh6JnBaeXpzMlbYTha2517eQmJcAWiZMVAAeTIAK1giT31DUFxYbHxcDGbW+GFwNGVYClkFJVVCiamrg6MjAFs7h58t3uFH8RgC-iCdAYTFY7Ag-woAEEvmk0GhnLx6HYIcF6IwWGwOECHgAhFHjXAYvBYnEhMAAOTQkWYsFatEoHASpSSHEkaHwUVikRqkAAbhxahw7LBGrlQPRQLVsqAAA5oeC0HKYagoaaIXrYfqNG4Ip6KGWvA1GwbDUaeE3A2R6SHBdBYEqYQ0DXBqEjEpFrEymLie61Lf2I6TmTz+N1exq+3ARyx-YFrPTpdGgGIUmLg108D19b2J-1k8RBkPF41lqMxnWhkt+hEklMI9OKmq0eBheCIijwLM5yAALxioHzdNAAHUSNBbPyVPLEMQ5PIVRRcpqKJJUMdshxkLrPukmnyzqAfraLpA6pBlFO0MxOxLaD8aMe2lX40xE8jnha+qNo0QwjBgAgRnW+gTjqerfmGiZkgBKhAdWNpgRBLZQc6BbuqAphzgumjKFaTZJgiyIVmY8FkYi0b+ARuDzrgi4kcBv7NsCiJtmm4h6KiWYAESCTBQA) and it can work

- [Recusive definitions](https://github.com/microsoft/TypeScript/issues/29872)
  
  You can write:
  ```ts
  export class Class1 {
    another: Class2
  }
  export class Class2 {
    another: Class1
  }
  ```
  But not when it's a mixin pattern:
  
  ```ts
    export const SampleMixin1 = <T extends AnyConstructor<object>>(base : T) =>
    class SampleMixin1 extends base {
        another: SampleMixin2 // TS2502: 'another' is referenced directly or indirectly in its own type annotation
    }
    export type SampleMixin1 = Mixin<typeof SampleMixin1>


    export const SampleMixin2 = <T extends AnyConstructor<object>>(base : T) =>
    class SampleMixin2 extends base {
        another: SampleMixin1 // TS2502: 'another' is referenced directly or indirectly in its own type annotation
    }
    export type SampleMixin2 = Mixin<typeof SampleMixin2>


    // supporting declarations for mixin pattern
    export type AnyFunction<A = any>      = (...input: any[]) => A
    export type AnyConstructor<A = any>   = new (...input: any[]) => A
    export type Mixin<T extends AnyFunction> = InstanceType<ReturnType<T>>
  ```
  
  Workaround seems to be to create an `interface` which extends the `typeof` for the class and use that. This isn't something we love, but it makes sense.

- [Mixins can't be represented in a d.ts file](https://github.com/microsoft/TypeScript/issues/35822) 

   - We can do it some tricky things to show some of these in .dts file
   - Orta will find some working examples and include them in the Docs. Look at the DTS output for [api-extractor](https://api-extractor.com) uses mixins, and can compare their dts.
 
- [Mixins can't be made with decorators](https://github.com/microsoft/TypeScript/issues/4881)

  Basically a decorator on a class can't mutate the shape of the class, though ts-mixer seems to do this (and you can use `interface` merging too)

## Notes

The pattern is supported, and used by the Vue community in v1/v2. We think mostly by the mixin factory method.

- There's a  [stage 1 proposal in tc39](https://github.com/justinfagnani/proposal-mixins), which is syntax sugar for the mixin factory pattern
- Maybe we should include a note on the variadic mixins
## For orta

-  look at the abstract constructor #36392 - which could improve mixin support too.

**Comment by orta** [comment]
I found both the issue on recursive mixing referencing and the DTS output to not really be an issue with the class expression pattern and so I've not added them to the constraints

**Comment by trusktr** [comment]
Does this issue consider the use of `protected` and `private` within class-factory mixins? I recall that the type workarounds don't work because mapped types can not handle protected/private properties.
